File: firebase-auth/src/main/java/com/androidhuman/rxfirebase2/auth/RxPhoneAuthProvider.java
package com.androidhuman.rxfirebase2.auth;

import com.google.firebase.auth.PhoneAuthProvider;

import android.app.Activity;
import android.support.annotation.CheckResult;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import java.util.concurrent.Executor;
import java.util.concurrent.TimeUnit;

import io.reactivex.Observable;

public final class RxPhoneAuthProvider {

    @CheckResult
    @NonNull
    public static Observable<PhoneAuthEvent> verifyPhoneNumber(
            @NonNull PhoneAuthProvider provider, String phoneNumber,
            long timeout, TimeUnit timeUnit, Activity activity) {
        return verifyPhoneNumber(provider, phoneNumber,
                timeout, timeUnit, activity, null);
    }

    @CheckResult
    @NonNull
    public static Observable<PhoneAuthEvent> verifyPhoneNumber(
            @NonNull PhoneAuthProvider provider, String phoneNumber,
            long timeout, TimeUnit timeUnit, Activity activity,
            @Nullable PhoneAuthProvider.ForceResendingToken forceResendingToken) {
        return new PhoneAuthProviderVerifyPhoneNumberActivityObserver(
                provider, phoneNumber, timeout, timeUnit, activity, forceResendingToken);
    }

    @CheckResult
    @NonNull
    public static Observable<PhoneAuthEvent> verifyPhoneNumber(
            @NonNull PhoneAuthProvider provider, String phoneNumber,
            long timeout, TimeUnit timeUnit, Executor executor) {
        return verifyPhoneNumber(provider, phoneNumber,
                timeout, timeUnit, executor, null);
    }

    @CheckResult
    @NonNull
    public static Observable<PhoneAuthEvent> verifyPhoneNumber(
            @NonNull PhoneAuthProvider provider, String phoneNumber,
            long timeout, TimeUnit timeUnit, Executor executor,
            @Nullable PhoneAuthProvider.ForceResendingToken forceResendingToken) {
        return new PhoneAuthProviderVerifyPhoneNumberExecutorObserver(
                provider, phoneNumber, timeout, timeUnit, executor, forceResendingToken);
    }
}

File: firebase-auth-kotlin/src/main/kotlin/com/androidhuman/rxfirebase2/auth/RxPhoneAuthProvider.kt
@file:Suppress("NOTHING_TO_INLINE", "UNUSED")

package com.androidhuman.rxfirebase2.auth

import android.app.Activity
import com.google.firebase.auth.PhoneAuthProvider
import io.reactivex.Observable
import java.util.concurrent.Executor
import java.util.concurrent.TimeUnit

inline fun PhoneAuthProvider.rxVerifyPhoneNumber(phoneNumber: String,
        timeout: Long, timeUnit: TimeUnit, activity: Activity)
        : Observable<PhoneAuthEvent>
        = RxPhoneAuthProvider.verifyPhoneNumber(
        this, phoneNumber, timeout, timeUnit, activity)

inline fun PhoneAuthProvider.rxVerifyPhoneNumber(phoneNumber: String,
        timeout: Long, timeUnit: TimeUnit, activity: Activity,
        forceResendingToken: PhoneAuthProvider.ForceResendingToken)
        : Observable<PhoneAuthEvent>
        = RxPhoneAuthProvider.verifyPhoneNumber(
        this, phoneNumber, timeout, timeUnit, activity, forceResendingToken)

inline fun PhoneAuthProvider.rxVerifyPhoneNumber(phoneNumber: String,
        timeout: Long, timeUnit: TimeUnit, executor: Executor)
        : Observable<PhoneAuthEvent>
        = RxPhoneAuthProvider.verifyPhoneNumber(
        this, phoneNumber, timeout, timeUnit, executor)

inline fun PhoneAuthProvider.rxVerifyPhoneNumber(phoneNumber: String,
        timeout: Long, timeUnit: TimeUnit, executor: Executor,
        forceResendingToken: PhoneAuthProvider.ForceResendingToken)
        : Observable<PhoneAuthEvent>
        = RxPhoneAuthProvider.verifyPhoneNumber(
        this, phoneNumber, timeout, timeUnit, executor, forceResendingToken)
